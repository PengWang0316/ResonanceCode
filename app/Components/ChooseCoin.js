import React, { Component } from "react";
import PropTypes from "prop-types";
// import $ from "jQuery";

class ChooseCoin extends Component{

  componentWillMount(){
    // setup the image file and put it in state
    this.headsImg=require("../imgs/heads.gif");
    this.tailsImg=require("../imgs/tails.gif");
    this.state={
      0: [this.headsImg,"-",false], //[image, title message, have content yet]
      1: [this.headsImg,"-",false],
      2: [this.headsImg,"-",false]
    }
    // setup the inintial value array for coins
    this.coinsArray=[2,2,2];
    // this.isHeadsArray=[true]
    // console.log(this.props.lineNumber,this.props.isAvailable);
  }

  onClickCoin(coinNum){
    if(this.coinsArray[coinNum]===3){
      this.coinsArray[coinNum]=2;
      this.state[coinNum][0]=this.tailsImg;
      this.state[coinNum][1]="T";
    }else{
      this.coinsArray[coinNum]=3;
      this.state[coinNum][0]=this.headsImg;
      this.state[coinNum][1]="H";
      this.state[coinNum][2]=true;
    }
    this.setState(this.state);
    // console.log(this.state);

  }

  handleSubmit(){
    let totalNum=0;
    this.coinsArray.map((element)=>{
      totalNum+=element;
    });
    // prepare heads and tails text
    let headsTails=`${this.state["0"][1]} ${this.state["1"][1]} ${this.state["2"][1]}`;
    this.props.handleCoinClick(this.props.lineNumber, totalNum, headsTails);
    this.props.handleCancel();
  }

  // handleCancel(){
  //   this.props.handleCancel();
  // }

  render(){

    return(
      <div className="coinPickUpBgDiv">
        <div className="coinPickUpWindowDiv">

          <div><b>Coins for line {this.props.lineNumber*1+1}</b></div>
          <div className="container">
            <div className="row text-center">
              <div className="col">
                <div onClick={()=>{this.onClickCoin(0)}}>{this.state["0"][2] ? <img src={this.state["0"][0]} className="coinImg" alt="Coin" title="Click this coin" /> : <div className="emptyCoinDiv">Click</div>}</div>
                <div id="coin_0">{this.state["0"][1]}</div>
              </div>
              <div className="col">
                <div onClick={()=>{this.onClickCoin(1)}}>{this.state["1"][2] ? <img src={this.state["1"][0]} className="coinImg" alt="Coin" title="Click this coin" /> : <div className="emptyCoinDiv">Click</div>}</div>
                <div id="coin_1">{this.state["1"][1]}</div>
              </div>
              <div className="col">
                <div onClick={()=>{this.onClickCoin(2)}}>{this.state["2"][2] ? <img src={this.state["2"][0]} className="coinImg" alt="Coin" title="Click this coin" /> : <div className="emptyCoinDiv">Click</div>}</div>
                <div id="coin_2">{this.state["2"][1]}</div>
              </div>
            </div>
          </div>


          <div className="text-right"><button type="button" className="btn btn-primary" disabled={!this.state["0"][2] || !this.state["1"][2] || !this.state["2"][2]} onClick={()=>{this.handleSubmit();}}>Submit</button><button type="button" className="btn btn-info" onClick={()=>{this.props.handleCancel();}}>Cancel</button></div>

        </div>
      </div>
    );
  }

}
ChooseCoin.propTypes={
  lineNumber: PropTypes.string.isRequired,
  handleCoinClick: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired
};
export default ChooseCoin;
