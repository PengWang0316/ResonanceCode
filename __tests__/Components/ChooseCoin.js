import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import ChooseCoin from '../../app/Components/ChooseCoin';

describe('ChooseCoin Test', () => {
  const props = {
    lineNumber: '1',
    handleCoinClick: jest.fn(),
    handleCancel: jest.fn(),
    handleSwitchMode: jest.fn()
  };
  const getShallowComponent = (newProps = props) => shallow(<ChooseCoin {...newProps} />);

  test('handleCancel ', () => {
    const handleCancel = jest.fn();
    const chooseCoin = getShallowComponent({ ...props, handleCancel });
    chooseCoin.find('button').at(1).simulate('click');
    expect(handleCancel).toHaveBeenCalledTimes(1);
  });

  test('handleSubmit', () => {
    const handleCancel = jest.fn();
    const chooseCoin = getShallowComponent({ ...props, handleCancel });
    chooseCoin.find('button').at(0).simulate('click');
    expect(props.handleCoinClick).toHaveBeenCalled();
    expect(handleCancel).toHaveBeenCalledTimes(1);
  });

  test('handleClickThirdCoin', () => {
    const chooseCoin = getShallowComponent();
    const button = chooseCoin.find({ role: 'button' }).at(3);
    button.simulate('click');
    expect(chooseCoin.state('2')[1]).toBe('H');
    expect(chooseCoin.state('2')[2]).toBe(true);
    button.simulate('click');
    expect(chooseCoin.state('2')[1]).toBe('T');
  });

  test('handleClickSecondCoin ', () => {
    const chooseCoin = getShallowComponent();
    const button = chooseCoin.find({ role: 'button' }).at(2);
    button.simulate('click');
    expect(chooseCoin.state('1')[1]).toBe('H');
    expect(chooseCoin.state('1')[2]).toBe(true);
    button.simulate('click');
    expect(chooseCoin.state('1')[1]).toBe('T');
  });

  test('handleClickFirstCoin  ', () => {
    const chooseCoin = getShallowComponent();
    const button = chooseCoin.find({ role: 'button' }).at(1);
    button.simulate('click');
    expect(chooseCoin.state('0')[1]).toBe('H');
    expect(chooseCoin.state('0')[2]).toBe(true);
    button.simulate('click');
    expect(chooseCoin.state('0')[1]).toBe('T');
  });

  test('button disable status', () => {
    const chooseCoin = getShallowComponent();
    expect(chooseCoin.find('button').at(0).prop('disabled')).toBe(true);
    chooseCoin.setState({
      0: [0, '-', true]
    });
    expect(chooseCoin.find('button').at(0).prop('disabled')).toBe(true);
    chooseCoin.setState({
      1: [1, '-', true]
    });
    expect(chooseCoin.find('button').at(0).prop('disabled')).toBe(true);
    chooseCoin.setState({
      2: [2, '-', true]
    });
    expect(chooseCoin.find('button').at(0).prop('disabled')).toBe(false);
  });

  test('ChooseCoin snapshot', () => expect(renderer.create(<ChooseCoin {...props} />).toJSON()).toMatchSnapshot());
});
