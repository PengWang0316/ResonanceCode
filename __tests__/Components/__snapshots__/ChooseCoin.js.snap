// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseCoin Test ChooseCoin snapshot 1`] = `
<div
  className="coinPickUpBgDiv"
>
  <div
    className="coinPickUpWindowDiv"
  >
    <div
      className="row mb-2"
    >
      <div
        className="col"
      >
        <b>
          Line 
          2
        </b>
      </div>
      <div
        className="btn-group btn-group-toggle col"
        data-toggle="buttons"
      >
        <label
          className="btn btn-secondary btn-sm active"
          htmlFor="coinMode"
        >
          <input
            autoComplete="off"
            defaultChecked={true}
            id="coinMode"
            name="options"
            type="radio"
          />
           Coin Mode
        </label>
        <label
          className="btn btn-outline-secondary btn-sm"
          htmlFor="HexagramMode"
          onClick={[MockFunction]}
          role="button"
          tabIndex="-1"
        >
          <input
            autoComplete="off"
            id="HexagramMode"
            name="options"
            type="radio"
          />
           Line Mode
        </label>
      </div>
    </div>
    <div
      className="container"
    >
      <div
        className="row text-center"
      >
        <div
          className="col"
        >
          <div
            onClick={[Function]}
            role="button"
            tabIndex="-2"
          >
            <div
              className="emptyCoinDiv"
            >
              Click
            </div>
          </div>
          <div
            id="coin_0"
          >
            -
          </div>
        </div>
        <div
          className="col"
        >
          <div
            onClick={[Function]}
            role="button"
            tabIndex="-1"
          >
            <div
              className="emptyCoinDiv"
            >
              Click
            </div>
          </div>
          <div
            id="coin_1"
          >
            -
          </div>
        </div>
        <div
          className="col"
        >
          <div
            onClick={[Function]}
            role="button"
            tabIndex="0"
          >
            <div
              className="emptyCoinDiv"
            >
              Click
            </div>
          </div>
          <div
            id="coin_2"
          >
            -
          </div>
        </div>
      </div>
    </div>
    <div
      className="text-right"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
      <button
        className="btn btn-info"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
